// @mml-io/observable-dom-common/build/index.js
// src/messages.ts
var ADD_CONNECTED_USER_ID_MESSAGE_TYPE = "addConnectedUserId";
var REMOVE_CONNECTED_USER_ID_MESSAGE_TYPE = "removeConnectedUserId";
var DISPATCH_REMOTE_EVENT_FROM_CONNECTION_ID_MESSAGE_TYPE = "dispatchRemoteEventFromConnectionId";
var DOM_MESSAGE_TYPE = "dom";
function applyMessageToObservableDOMInstance(message, instance) {
  if (message.type === ADD_CONNECTED_USER_ID_MESSAGE_TYPE) {
    instance.addConnectedUserId(message.connectionId, message.connectionToken);
  } else if (message.type === REMOVE_CONNECTED_USER_ID_MESSAGE_TYPE) {
    instance.removeConnectedUserId(message.connectionId);
  } else if (message.type === DISPATCH_REMOTE_EVENT_FROM_CONNECTION_ID_MESSAGE_TYPE) {
    instance.dispatchRemoteEventFromConnectionId(message.connectionId, message.event);
  } else {
    console.error("Unknown message type", message);
  }
}
function observableDOMInterfaceToMessageSender(sender, dispose) {
  const remoteObservableDOM = {
    addConnectedUserId(connectionId, connectionToken) {
      sender({
        type: ADD_CONNECTED_USER_ID_MESSAGE_TYPE,
        connectionId,
        connectionToken
      });
    },
    dispatchRemoteEventFromConnectionId(connectionId, remoteEvent) {
      sender({
        type: DISPATCH_REMOTE_EVENT_FROM_CONNECTION_ID_MESSAGE_TYPE,
        connectionId,
        event: remoteEvent
      });
    },
    dispose() {
      dispose();
    },
    removeConnectedUserId(connectionId) {
      sender({
        type: REMOVE_CONNECTED_USER_ID_MESSAGE_TYPE,
        connectionId
      });
    }
  };
  return remoteObservableDOM;
}
export {
  ADD_CONNECTED_USER_ID_MESSAGE_TYPE,
  DISPATCH_REMOTE_EVENT_FROM_CONNECTION_ID_MESSAGE_TYPE,
  DOM_MESSAGE_TYPE,
  REMOVE_CONNECTED_USER_ID_MESSAGE_TYPE,
  applyMessageToObservableDOMInstance,
  observableDOMInterfaceToMessageSender
};
